### PR TITLE
Add congestionControl contructor arg and readonly attribute

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes https://github.com/w3c/webtransport/issues/365.

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/webtransport/pull/406.html" title="Last updated on Aug 30, 2022, 2:20 PM UTC (3f4b036)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webtransport/406/71ee094...jan-ivar:3f4b036.html" title="Last updated on Aug 30, 2022, 2:20 PM UTC (3f4b036)">Diff</a>

### CODE REVIEW

**Comment by jan-ivar** [comment]
Meeting:
- Are the abstract enum values sufficient to pick between multiple congestion control algorithms, or would it be preferable to accept a string with name(s)? NewReno, BBR, SCReAM etc.
- Might need a registry approach instead
- note: getter is usable for feature detection immediately, i.e. this isn't a negotiated property dependent on server
- might be good to have both abstract ones and specific names.
- modify PR to propose a static getter getAvailableAlgorithms.

**Comment by LPardue** [comment]
> If the user agent only implements the default congestion control algorithm in
   [[!RFC9002]] [section 7](https://datatracker.ietf.org/doc/html/rfc9002#section-7),
   set |congestionControl| to `"throughput"`.

> This configuration option is considered at risk due to the lack of
       implementation and specification of any congestion control algorithm
       other than [[!RFC9002]] [section 7](https://datatracker.ietf.org/doc/html/rfc9002#section-7)
       at the time of writing.
       
I don't think these statements are quite accurate. The Reno-like CC in RFC 9002 is exemplary, IIUC  most implementations don't use it and prefer CUBIC - including Firefox and Chrome.

File: index.bs [hunk 1]
```diff
@@ -638,6 +645,11 @@ agent MUST run the following steps:
    {{TypeError}}.
 1. Let |requireUnreliable| be {{WebTransport/constructor(url, options)/options}}'s
    {{WebTransportOptions/requireUnreliable}}.
+1. Let |congestionControl| be {{WebTransport/constructor(url, options)/options}}'s
+   {{WebTransportOptions/congestionControl}}.
+1. If the user agent only implements the default congestion control algorithm in
+   [[!RFC9002]] [section 7](https://datatracker.ietf.org/doc/html/rfc9002#section-7),
```
Comment (martinthomson) [inline index.bs:651]: that's fairly narrowly targeted, so maybe "... or an algorithm that optimizes for throughput ..."

File: index.bs [hunk 2]
```diff
@@ -1058,6 +1085,18 @@ that determine how WebTransport connection is established and used.
    use for normal [=fetch=] operations.
 :: This cannot be used with {{WebTransportOptions/allowPooling}}.
 
+: <dfn for="WebTransportOptions" dict-member>congestionControl</dfn>
+:: When set, influences the congestion control algorithm used by the user agent
+   when sending data over this connection.
+   <div class="issue atrisk">
```
Comment (martinthomson) [inline index.bs:1091]: 👍

File: index.bs [hunk 3]
```diff
@@ -1058,6 +1085,18 @@ that determine how WebTransport connection is established and used.
...
+     <p>
+       This configuration option is considered at risk due to the lack of
+       implementation and specification of any congestion control algorithm
+       other than [[!RFC9002]] [section 7](https://datatracker.ietf.org/doc/html/rfc9002#section-7)
+       at the time of writing.
```
Comment (martinthomson) [inline index.bs:1096]: What Lucas said.  The challenge here is in standardizing an algorithm that optimizes for latency.  So maybe say that instead.

**Comment by jan-ivar** [comment]
OK I've reworked the API based on meeting feedback, with some liberties. Example usage:

Read which CC a connection uses (readonly):
```js
const wt = new WebTransport(url);
console.log(wt.congestionControl.name); // CUBIC
console.log(wt.congestionControl.aim); // throughput
```

Read the CC algorithms the browser supports, and create a WebTransport with the first one:
```js
const [congestionControl] = WebTransport.getCongestionControlAlgorithms();
console.log(JSON.stringify(congestionControl)); // {name: "CUBIC", aim: "throughput"}
const wt = new WebTransport(url, {congestionControl}); // always works
```
Feature-detection:
```js
if (!WebTransport.getCongestionControlAlgorithms().find(({aim}) => aim == "low-latency")) {
  console.log("Requires low-latency");
}
```
```js
if (!WebTransport.getCongestionControlAlgorithms().find(({name}) => name == "SCReAM")) {
  console.log("Requires SCReAM");
}
```

Probing with the constructor also works (throws if not supported):
```js
try {
  const wt = new WebTransport(url, {congestionControl: {aim: "low-latency"});
} catch (e) {
  if (e.name == "NotSupportedError") console.log("Requires low-latency");
  throw e;
}
```
```js
try {
  const wt = new WebTransport(url, {congestionControl: {name: "SCReAM"});
} catch (e) {
  if (e.name == "NotSupportedError") console.log("Requires SCReAM");
  throw e;
}
```
LMKWYT

**Comment by martinthomson** [comment]
There was a running joke during the later stages of QUIC standardization.  It goes something like this:

> Given two choices, we will always choose the most complex one.

It sounds like this is continuing in that fine tradition.

**Comment by jan-ivar** [comment]
Would a way forward be to go over the requirements, get the WG to agree what they are, and use that to pick?

The tension here seems to be how low-level we want this API to be, and how much abstraction such apps can tolerate, vs how much responsibility we expect them to handle.

One litmus question might be: if two browsers end up defaulting to a different CC for `"throughput"`, would that break apps?

**Comment by jan-ivar** [comment]
I have an action to update this PR based on https://github.com/w3c/webtransport/issues/365#issuecomment-1202704563.

**Comment by jan-ivar** [comment]
Note this PR deviates from https://github.com/w3c/webtransport/issues/365#issuecomment-1202704563 in one regard: the constructor takes one of 3 enum values: `"default"`, `"throughput"` or `"low-latency"`.

This seems necessary to let an application truly specify whether it values throughput or low-latency more.

If the choices were only between `"default"` and `"low-latency"`, then my concern is the user agent would have no way to detect an application that values throughput, unless it interprets `"default"` to mean this, which risks cementing throughput as the default, which seems undesirable.

**Comment by jan-ivar** [comment]
Meeting:
- General agreement on this direction
- Preference expressed for throwing enum
- Preference for the readable property to be more useful: a value other than default signifies that you got what you wanted
- Craft some prose to let the user agent decide

File: index.bs [hunk 4]
```diff
@@ -1058,6 +1091,18 @@ that determine how WebTransport connection is established and used.
...
 :: This cannot be used with {{WebTransportOptions/allowPooling}}.
 
+: <dfn for="WebTransportOptions" dict-member>congestionControl</dfn>
+:: Optionally specifies an application's preference for a congestion control
+   algorithm tuned for either throughput or low-latency to be used when sending
+   data over this connection. This is a hint to the user agent.
+   <div class="issue atrisk">
+     <p>
+       This configuration option is considered a feature at risk due to the lack
+       of standardization and implementation of a congestion control algorithm,
```
Comment (jan-ivar) [inline index.bs:1101]:
```suggestion
       of implementation of a congestion control algorithm,
```

**Pull request merged**